src/com/google/enterprise/adaptor/database/DatabaseAdaptor.java

```diff
@@ -262,18 +263,19 @@ public void getDocIds(DocIdPusher pusher) throws IOException,
    * Adds all specified metadata columns to the {@code DocIdPusher.Record} being
    * built.
    */
+  // TODO(bmj): This will throw SQLException if one of the columnNames in the
+  // Map does not exist in the ResultSet. We should think about how to deal with
+  // that, especially in getDocIds().
   @VisibleForTesting
   void addMetadataToRecordBuilder(DocIdPusher.Record.Builder builder,
       ResultSet rs) throws SQLException {
     ResultSetMetaData rsMetaData = rs.getMetaData();
-    int numberOfColumns = rsMetaData.getColumnCount();
-    for (int i = 1; i < (numberOfColumns + 1); i++) {
-      String columnName = rsMetaData.getColumnLabel(i);
-      Object value = rs.getObject(i);
-      String key = metadataColumns.getMetadataName(columnName);
-      if (key != null) {
-        builder.addMetadata(key, "" + value);
-      }
+    MetadataColumns metadataCols = (metadataColumns == null)
```

Comment:
As discussed, ponder a synchronized block to lazy-initialize a null metadataColumns field here.